src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs

```diff
@@ -389,126 +390,114 @@ public KestrelHttpParser(IKestrelTrace log)
                     break;
                 }
-
                 if (ch == ByteTab || ch == ByteSpace)
                 {
                     sawWhitespace = true;
                 }
+            }
 
-                index++;
-                pCurrent++;
+            if (index == length)
+            {
+                RejectRequest(RequestRejectionReason.NoColonCharacterFoundInHeaderLine);
             }
-            return result;
+            if (sawWhitespace)
+            {
+                RejectRequest(RequestRejectionReason.WhitespaceIsNotAllowedInHeaderName);
+            }
+            return index;
         }
 
-        private unsafe void TakeSingleHeader<T>(byte* pHeader, int headerLineLength, T handler) where T : IHttpHeadersHandler
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void TakeSingleHeader<T>(byte* headerLine, int length, T handler) where T : IHttpHeadersHandler
         {
-            var nameEnd = -1;
-            var valueStart = -1;
-            var valueEnd = -1;
-            var index = 0;
-            var pCurrent = pHeader + index;
-            var pEnd = pHeader + headerLineLength;
-
-            nameEnd = IndexOfNameEnd(pHeader, index, headerLineLength);
+            // Skip CR, LF from end position
+            var valueEnd = length - 3;
+            var nameEnd = IndexOfNameEnd(headerLine, length);
 
-            if (nameEnd == -1)
+            if (headerLine[valueEnd + 2] != ByteLF)
             {
-                RejectRequest(RequestRejectionReason.NoColonCharacterFoundInHeaderLine);
+                goto headerValueContainsCR;
+            }
+            if (headerLine[valueEnd + 1] != ByteCR)
+            {
+                RejectRequest(RequestRejectionReason.MissingCRInHeaderLine);
             }
 
-            // Skip colon
-            index += nameEnd + 1;
-            pCurrent += index;
-            valueStart = index;
-            var pValueStart = pCurrent;
-
-            while (pCurrent < pEnd)
+            // Skip colon from value start
+            var valueStart = nameEnd + 1;
+            // Ignore start whitespace
+            for(; valueStart < valueEnd; valueStart++)
             {
-                var ch = *pCurrent;
+                var ch = headerLine[valueStart];
                 if (ch != ByteTab && ch != ByteSpace && ch != ByteCR)
                 {
-                    valueStart = index;
-                    pValueStart = pCurrent;
                     break;
                 }
                 else if (ch == ByteCR)
                 {
-                    break;
+                    goto headerValueContainsCR;
                 }
-                pCurrent++;
-                index++;
             }
 
-            var endIndex = headerLineLength - 1;
-            pEnd--;
 
-            if (*pEnd != ByteLF)
+            // Check for CR in value
+            var i = valueStart + 1;
+            if (Vector.IsHardwareAccelerated)
```

Comment:
Ah, sorry, my bad, it looked like you hand inlined IndexOf, can we move this to `bool Contains(Span<byte>, byte)` method?